WcaOnRails/spec/requests/registrations_spec.rb

```diff
@@ -580,133 +584,256 @@
           post registration_payment_intent_path(registration.id), params: {
-            payment_method_id: pm.id,
             amount: registration.outstanding_entry_fees.cents,
           }
-          expect_error_to_be(response, "Your card's security code is incorrect.")
+          payment_intent = registration.reload.stripe_payment_intents.first
+
+          expect {
+            # mimic the user clicking through the interface
+            Stripe::PaymentIntent.confirm(
+              payment_intent.stripe_id,
+              { payment_method: 'pm_card_visa_chargeDeclinedIncorrectCvc' },
+              stripe_account: competition.connected_stripe_account_id,
+            )
+          }.to raise_error(Stripe::StripeError, "Your card's security code is incorrect.")
+
+          expect {
+            # mimick the response that Stripe sends to our return_url after completing the checkout UI
+            get registration_payment_success_path(registration.id), params: {
+              payment_intent: payment_intent.stripe_id,
+              payment_intent_client_secret: payment_intent.client_secret,
+            }
+          }.to_not change { registration.reload.outstanding_entry_fees }
+
+          expect(registration.paid_entry_fees).to eq 0
+          expect(payment_intent.confirmed_at).to be_nil
+          expect(payment_intent.stripe_transaction.reload.status).to eq('requires_payment_method')
+          expect(payment_intent.stripe_transaction.error).to eq('incorrect_cvc')
         end
 
         it "records a failure in the stripe journal" do
-          expect(StripeTransaction.all.length).to eq 0
-          card = FactoryBot.create(:credit_card, :invalid)
-          pm = Stripe::PaymentMethod.create(
-            { type: "card", card: card },
-            stripe_account: competition.connected_stripe_account_id,
-          )
+          expect(StripeTransaction.count).to eq 0
+          expect(StripePaymentIntent.count).to eq 0
+
           post registration_payment_intent_path(registration.id), params: {
-            payment_method_id: pm.id,
             amount: registration.outstanding_entry_fees.cents,
           }
-          stripe_charge = StripeTransaction.first
-          expect(stripe_charge&.status).to eq "failure"
-          metadata = stripe_charge.parameters
-          expect(metadata["payment_method"]).to eq pm.id
-          expect(metadata["metadata"]["competition"]).to eq competition.name
+          payment_intent = registration.reload.stripe_payment_intents.first
+          expect(payment_intent).to_not be_nil
+          # Intent should not be confirmed at this stage, because we have never received a receipt charge from Stripe yet
+          expect(payment_intent.confirmed_at).to be_nil
+
+          expect {
+            # mimic the user clicking through the interface
+            Stripe::PaymentIntent.confirm(
+              payment_intent.stripe_id,
+              { payment_method: 'pm_card_visa_chargeDeclined' },
```

Comment:
Would it be possible to add tests for other payment methods too?

Reply:
Absolutely!

Reply:
...or not. Stripe does not provide convenient PaymentMethod ID shorthands for payment methods other than credit cards.
Manually creating such PaymentMethod objects is possible but very cumbersome because it requires pre-configuring the PI and also adding stuff like `mandate_data` for processing SEPA payment methods etc...